src/mlang/m_ir/mir.ml

```diff
@@ -411,6 +411,19 @@ let find_var_by_name (p : program) (name : string Pos.marked) : Variable.t =
            (Pos.VarNameToID.find name p.program_idmap))
     with Not_found -> Errors.raise_spanned_error "unknown variable" (Pos.get_position name))
 
+let find_vars_by_io (p : program) (io_to_find : io) : VariableDict.t =
+  let rules_as_list = RuleMap.bindings p.program_rules in
+  let rule_vars =
+    List.flatten (List.map (fun rd -> rd.rule_vars) (List.map (fun (_, rd) -> rd) rules_as_list))
```

Comment:
Since we're now in 4.11, we have `List.concat_map` that combine `flatten` and `map` in an efficient (and tail-recursive !) way.
Also `(fun (_, rd) -> rd)` is effectively the pervasive function `snd`.